[Android] RTE ShellFragmentContainer

### Description

Navigating from the home page to favorites in WeatherTwentyOne on Android 12 I crash:

```
[mono-rt] [ERROR] FATAL UNHANDLED EXCEPTION: System.ObjectDisposedException: Cannot access a disposed object.
[mono-rt] Object name: 'Microsoft.Maui.Controls.Platform.Compatibility.ShellFragmentContainer'.
[mono-rt]    at Java.Interop.JniPeerMembers.AssertSelf(IJavaPeerable self) in /Users/runner/work/1/s/xamarin-android/external/Java.Interop/src/Java.Interop/Java.Interop/JniPeerMembers.cs:line 152
[mono-rt]    at Java.Interop.JniPeerMembers.JniInstanceMethods.InvokeVirtualVoidMethod(String encodedMember, IJavaPeerable self, JniArgumentValue* parameters) in /Users/runner/work/1/s/xamarin-android/external/Java.Interop/src/Java.Interop/Java.Interop/JniPeerMembers.JniInstanceMethods_Invoke.cs:line 24
[mono-rt]    at AndroidX.Fragment.App.Fragment.OnDestroy() in D:\a\1\s\generated\androidx.fragment.fragment\obj\Release
et6.0-android\generated\src\AndroidX.Fragment.App.Fragment.cs:line 2023
[mono-rt]    at Microsoft.Maui.Controls.Platform.Compatibility.ShellFragmentContainer.OnDestroy() in D:\a\_work\1\s\src\Compatibility\Core\src\Handlers\Shell\Android\ShellFragmentContainer.cs:line 45
[mono-rt]    at AndroidX.Fragment.App.Fragment.n_OnDestroy(IntPtr jnienv, IntPtr native__this) in D:\a\1\s\generated\androidx.fragment.fragment\obj\Release
et6.0-android\generated\src\AndroidX.Fragment.App.Fragment.cs:line 2013
[mono-rt]    at Android.Runtime.JNINativeWrapper.Wrap_JniMarshal_PP_V(_JniMarshal_PP_V callback, IntPtr jnienv, IntPtr klazz) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.Runtime/JNINativeWrapper.g.cs:line 22
```

This works ok on Windows.

### Steps to Reproduce

Run https://github.com/davidortinau/WeatherTwentyOne on Android 12 and nav from the start page to Favorites.

### Version with bug

Release Candidate 1

### Last version that worked well

Preview 13 (current)

### Affected platforms

Android

### Affected platform versions

12

### Did you find any workaround?

_No response_

### Relevant log output

_No response_

--- a/src/Compatibility/Core/src/Handlers/Shell/Android/ShellFragmentContainer.cs
+++ b/src/Compatibility/Core/src/Handlers/Shell/Android/ShellFragmentContainer.cs
@@ -22,9 +22,9 @@ namespace Microsoft.Maui.Controls.Platform.Compatibility
 		public override AView OnCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
 		{
 			_page = ((IShellContentController)ShellContentTab).GetOrCreateContent();
-			
+
 			var pageMauiContext = _mauiContext.MakeScoped(layoutInflater: inflater, fragmentManager: ChildFragmentManager);
-			
+
 			return new ShellPageContainer(RequireContext(), (IPlatformViewHandler)_page.ToHandler(pageMauiContext), true)
 			{
 				LayoutParameters = new LP(LP.MatchParent, LP.MatchParent)
@@ -40,7 +40,9 @@ namespace Microsoft.Maui.Controls.Platform.Compatibility
 
 		public override void OnDestroy()
 		{
-			Activity.RunOnUiThread(Dispose);
+			_mauiContext
+				.GetDispatcher()
+				.Dispatch(Dispose);
 
 			base.OnDestroy();
 		}

--- a/src/Compatibility/Core/src/Handlers/Shell/Android/ShellItemRendererBase.cs
+++ b/src/Compatibility/Core/src/Handlers/Shell/Android/ShellItemRendererBase.cs
@@ -292,7 +292,7 @@ namespace Microsoft.Maui.Controls.Platform.Compatibility
 
 		protected virtual void OnShellSectionChanged()
 		{
-			HandleFragmentUpdate(ShellNavigationSource.ShellSectionChanged, ShellSection, null, false);
+			HandleFragmentUpdate(ShellNavigationSource.ShellSectionChanged, ShellSection, null, false).FireAndForget();
 		}
 
 		protected virtual void OnDisplayedPageChanged(Page newPage, Page oldPage)

--- a/src/Controls/tests/DeviceTests/Elements/Shell/ShellBasicNavigationTestCases.cs
+++ b/src/Controls/tests/DeviceTests/Elements/Shell/ShellBasicNavigationTestCases.cs
@@ -1,0 +1,52 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Maui.Controls;
+
+namespace Microsoft.Maui.DeviceTests
+{
+	public class ShellBasicNavigationTestCases : IEnumerable<object[]>
+	{
+		public IEnumerator<object[]> GetEnumerator()
+		{
+			var page1 = new ContentPage();
+			var page2 = new ContentPage();
+
+			yield return new object[] { new ShellItem[]
+				{
+					new ShellItem() { Items = { page1 }, Route = "page1" },
+					new ShellItem() { Items = { page2 }, Route = "page2" }
+				} };
+
+			yield return new object[] {new ShellItem[]
+				{
+					new ShellSection() { Items = { page1 }, Route = "page1" },
+					new ShellSection() { Items = { page2 }, Route = "page2" }
+				} };
+
+			yield return new object[] { new ShellItem[]
+				{
+					new ShellContent() { Content = page1, Route = "page1" },
+					new ShellContent() { Content = page2, Route = "page2" }
+				} };
+
+			yield return new object[] { new ShellItem[]
+				{
+					new FlyoutItem()
+					{
+						Items =
+						{
+							new ShellContent() { Content = page1, Route = "page1" },
+							new ShellContent() { Content = page2, Route = "page2" },
+							new ShellContent() { Content = new ContentPage(), Route = "page3" },
+						}
+					}
+				} };
+		}
+
+		IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+	}
+}

--- a/src/Controls/tests/DeviceTests/Elements/Shell/ShellTests.cs
+++ b/src/Controls/tests/DeviceTests/Elements/Shell/ShellTests.cs
@@ -50,20 +50,46 @@ namespace Microsoft.Maui.DeviceTests
 		{
 			SetupBuilder();
 
-			var shell = await InvokeOnMainThreadAsync<Shell>(() => {
+			var shell = await InvokeOnMainThreadAsync<Shell>(() =>
+			{
 				return new Shell()
-					{
-						Items =
+				{
+					Items =
 						{
 							new ContentPage()
 						}
-					};
+				};
 			});
 
-			await CreateHandlerAndAddToWindow<ShellHandler>(shell, (handler) =>
+			await CreateHandlerAndAddToWindow<ShellHandler>(shell, async (handler) =>
 			{
+				// TODO MAUI Fix this 
+				await Task.Delay(100);
 				Assert.NotNull(shell.Handler);
-				return Task.CompletedTask;
+			});
+		}
+
+
+		[Theory]
+		[ClassData(typeof(ShellBasicNavigationTestCases))]
+		public async Task BasicShellNavigationStructurePermutations(ShellItem[] shellItems)
+		{
+			SetupBuilder();
+			var shell = await InvokeOnMainThreadAsync<Shell>(() =>
+			{
+				var value = new Shell();
+				foreach (var item in shellItems)
+					value.Items.Add(item);
+
+				return value;
+			});
+
+			await CreateHandlerAndAddToWindow<ShellHandler>(shell, async (handler) =>
+			{
+				// TODO MAUI Fix this 
+				await Task.Delay(100);
+				await shell.GoToAsync("//page2");
+				await Task.Delay(100);
 			});
 		}
 #endif

